README.md
# react-widget-socketio-html

A Rasa socket.io channel supported React chat widget component attachable via &lt;script> tags. Only supports English.

## Setting up Dev Environmet

The original package was created using `npx create-react-app`. So, to set up a dev environment, the `node_modules` dir initialized using `npx create-react-app` is required to begin with.

- Clone this repo
- Simply run `npm run init` to initialize required node_modeules
- Run `npm i` to install additional packages
- Run `npm start` to run the dev server
- This project adhears to all the commands executable inside a `npx create-react-app` project
- Additionally, this project supports the following npm commands
    1. `npm run init` - initializes required node_modules (takes care of cleanup internally)
    2. `npm run cleanup` - cleans up cached project files
    3. `npm run release` - creates a release that is ready to be attached via a script tag and places it under the `release` dir under the specific version. (takes care of `npm run build` internally)

## Releasing

- When the version is bumped, first do the versioning in the package.json
- run `npm run build` to build the latest build. (This is optional. `npm run release` can take care of this internally)
- run `npm run release` to get the transpiled js module to `release` dir under the proper package version.
- In addition to the js module, a style sheet and a html document where everything is attached for testing are generated automatically.

## Attaching via a `<script>` tag

The component is attachable to any html via a `<script>` tag as follows. You may refer to the test HTML included in each release dir as well.

```html
<noscript>You need to enable JavaScript to run this app.</noscript>
<script>
  window.kolloqe_webchat_id = "testid";
  window.kolloqeProps = {
    testid: {
      position: "right",
      positionProps: {
        xAxis: "10px",
        yAxis: "10px"
      },
      title: "Kolloqe ⚡",
      subtitle: "chat widget",
      avatar: true,
      avatarType: "text",
      avatarLink: "",
      socketURL: "http://localhost:5005",
      initPayload: "/get_started",
      reconnectTimeout: 5000,
      widgetOptions: true,
      fullscreenButton: false,
      moreOptions: true,
      widgetOptionsPosition: "top",
      hideWhenNotConnected: true,
      displayUnreadCount: false,
      showMessageDate: false,
      customMessageDelay: null,
      persistSession: true,
      triggerIntents: true,
      enableURLs: true,
      disablePreviousQuickReplies: true,
      indicatorDelay: 0,
      indicatorType: "bouncing",
      params: null,
    }
  };

  const version = "1.0.0";
  const source = "kolloqe-widget-socketio-en-html-v1.0.0";

  !(function () {
    let e = document.createElement("script");
    let s = document.createElement("link");
    let t = document.head || document.getElementsByTagName("head")[0];
    (e.src = "kolloqe-widget-socketio-en-html-v1.0.0.js"),
      (e.async = !0),
      (e.onload = () => { }),
      t.insertBefore(e, t.firstChild);

    (s.href = "kolloqe-widget-socketio-en-html-v1.0.0.css"),
      (s.rel = "stylesheet"),
      (s.type = "text/css"),
      t.insertBefore(s, t.firstChild);
  })();
</script>
```

## Passing Props

{To be updated}
Refer the [Official Docs](https://kolloqe.github.io) for more details on passing props to kolloqe chat widget component.

## Limitations

- Overriding styles has to be done manually or required to rebuild a different version of the component.
- Does not support snapping to fullscreen yet in smaller screen sizes. We're working on that.


CHANGELOG.md
# react-widget-socketio-en-html Change Log

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](https://semver.org/) starting with version 1.0.0

## [1.0.1] - 2022-07-28

### Bugfixes

- fixed a bug in handling language shortcut enabling/disabling prop passing

## [1.0.0] - 2022-07-28

### Features

- Fully supports socket.io channels in Rasa bots (Rasa open-source version 2.8.x)
- Fully customizable including avatar, title, subtitle, message placeholder and more
- Customizable options pane attachable either on top or bottom of the chat widget
- Individual options in the option pane can be hidden via props
- Fully supports English. Check [Kolloqe](https://kolloqe.github.io) out which is a product that uses the chatwidget which has been extended through supporting code-switching between Sinhala (Si) and English (En)
- Supports rich responses including text, text + URLs, URLs, Images, Buttons (Quick Replies), and Cards (Carousels)
- Button behavior is fully configurable
- Offers support to specify various button types including Action buttons, Resonse buttons, Link buttons, and Dummy Buttons both in Quick Replies and Carousel buttons
- Clear chat ability
- Text-based avatar support
- Session persistence support

### Deprecations and Removals

- Dropped support for code-switching between Sinhala and English from the initial test version since it is a feature attached with Kolloqe, a propriotary conversational AI platform


src/index.js
import React from 'react';
import ReactDOM from 'react-dom/client';
import KolloqeChatWidget from './webchat/KolloqeChatWidget';

let element_id = undefined;
let root = undefined;
let props = {};

try {
  element_id = window?.kolloqe_webchat_id;
  props = window?.kolloqeProps[element_id];
  root = ReactDOM.createRoot(document.getElementById(String(element_id)));
} catch (err) {
  console.error("failed to retrieve the props");
}

root.render(
  <React.StrictMode>
    <div id="kolloqe-widget-container">
      <KolloqeChatWidget
        position={props?.position}
        positionProps={props?.positionProps}
        title={props?.title}
        subtitle={props?.subtitle}
        defaultLang={props?.defaultLang}
        avatar={props?.avatar}
        avatarType={props?.avatarType}
        avatarLink={props?.avatarLink}
        socketURL={props?.socketURL}
        initPayload={props?.initPayload}
        reconnectTimeout={props?.reconnectTimeout}
        widgetOptions={props?.widgetOptions}
        fullscreenButton={props?.fullscreenButton}
        langSwitch={props?.langSwitch}
        moreOptions={props?.moreOptions}
        widgetOptionsPosition={props?.widgetOptionsPosition}
        hideWhenNotConnected={props?.hideWhenNotConnected}
        displayUnreadCount={props?.displayUnreadCount}
        showMessageDate={props?.showMessageDate}
        persistSession={props?.persistSession}
        triggerIntents={props?.triggerIntents}
        enableURLs={props?.enableURLs}
        disablePreviousQuickReplies={props?.disablePreviousQuickReplies}
        indicatorDelay={props?.indicatorDelay}
        indicatorType={props?.indicatorType}
        enableShortcuts={props?.enableShortcuts}
        shortcutKey={props?.shortcutKey} />
    </div>
  </React.StrictMode>
);


scripts/init.js
const fs = require('fs');
const fse = require('fs-extra');
var rimraf = require("rimraf");
var exec = require('child_process').exec;

const cacheProjectPath = "cache";
const cacheNodeModulesPath = "cache/node_modules";
const destNodeModulesPath = "node_modules";

const dirExists = fs.existsSync(cacheNodeModulesPath);
if (!dirExists) {
  {
    console.log(
      "\x1b[32m",
      "Initializing a cache project...",
      "\x1b[0m"
    );
    var cmd = exec("npx create-react-app cache", function (err, stdout, stderr) {
      if (err) {
        console.error(
          "\x1b[31m",
          `Error occurred while initializing the cache project.`,
          "\x1b[0m"
        );
        console.error(err);
        return;
      } else {
        console.log(
          "\x1b[32m",
          "Successfully initialized the cache project.",
          "\x1b[0m"
        );
        triggerInit(dirExists, cacheNodeModulesPath, destNodeModulesPath, cacheProjectPath);
      }
      // console.log(stdout);
    });
  }
} else {
  triggerInit(dirExists, cacheNodeModulesPath, destNodeModulesPath, cacheProjectPath);
}

function triggerInit(dirExists, cacheNodeModulesPath, destNodeModulesPath, cacheProjectPath) {
  if (dirExists) {
    console.log(
      "\x1b[32m",
      "Copying required node_modules...",
      "\x1b[0m"
    );
    fse.copy(cacheNodeModulesPath, destNodeModulesPath, { overwrite: true })
      .then(() => {
        console.log(
          "\x1b[32m",
          `Successfully copied the required node_modules.`,
          "\x1b[0m"
        );
        // deleting the dummy react project
        try {
          console.log(
            "\x1b[32m",
            `Cleanup triggerred...`,
            "\x1b[0m"
          );
          rimraf.sync(cacheProjectPath);
          console.log(
            "\x1b[32m",
            `Successfully cleaned up the cached project.`,
            "\x1b[0m"
          );
        } catch (err) {
          console.error(
            "\x1b[31m",
            `Failed to remove cached project. Please manually remove 'cache' dir in the project root if it exists.`,
            "\x1b[0m"
          );
        }
      })
      .catch((err) => {
        console.error(
          "\x1b[31m",
          `Failed to copy the required node_modules.`,
          "\x1b[0m"
        );
        console.error(err);
      })
  } else {
    console.error(
      "\x1b[31m",
      `Failed to initialize the required node_modules.`,
      "\x1b[0m"
    );
  }
}

src/webchat/configs.js
export const localStorageKeys = {
  kolloqeWidgetOpen: "kolloqeWidgetOpen",
  kolloqeChatSession: "kolloqeChatSession",
  kolloqeLocalChatSession: "kolloqeLocalChatSession",
  kolloqeSelectedLang: "kolloqeSelectedLang"
}

export const webchatConfigs = {
  initPayload: "/get_started",
  urlEndpoint: "http://localhost:5005",
  socketPath: "/socket.io/",
  title: "Kolloqe ⚡",
  customData: {"language": "en"},
  docViewer: true,
  subtitle: null,
  inputTextFieldHint: "Type a message",
  hideWhenNotConnected: true,
  // connectOn
  // onSocketEvent: {
  //   'bot_uttered': () => console.log('the bot said something'),
  //   'connect': () => console.log('connection established'),
  //   'disconnect': () => doSomeCleanup(),
  // },
  embedded: false,
  showFullScreenButton: false,
  displayUnreadCount: false,
  showMessageDate: false,
  // customMessageDelay
  // params: {
  //   images: {
  //     dims: {
  //       width: 300,
  //       height: 200
  //     }
  //   }
  // },
  storage: "local",
  customComponent: null,
  onWidgetEvent: {},
  linkRegex: /(\[[a-zA-Z0-9\\,./{}!~#$%^&*()_\-+=:;'"`@<> ]+\]\([\S]+\))/g,
  subLinkRegex: /\[([a-zA-Z0-9\\,./{}!~#$%^&*()_\-+=:;'"`@<> ]+)\]\(([\S]+)\)/g,
}

scripts/release.js
const fs = require('fs');
const glob = require("glob");
const path = require("path");
var packageJSON = require('../package.json');

const buildFilePattern = "build/static/js/*.js";
const buildStylesPattern = "build/static/css/*.css";
const releaseDirPath = `release/v${packageJSON.version}/`;
const releaseFilePath = `release/v${packageJSON.version}/kolloqe-widget-socketio-en-html-v${packageJSON.version}.js`;
const releaseStylesPath = `release/v${packageJSON.version}/kolloqe-widget-socketio-en-html-v${packageJSON.version}.css`;
const releaseHTMLPath = `release/v${packageJSON.version}/kolloqe-widget-socketio-en-html-v${packageJSON.version}-test.html`;

const componentExists = fs.existsSync(releaseFilePath);
const styleExists = fs.existsSync(releaseStylesPath);

if (componentExists || styleExists) {
  log("Release already exists! Consider bumping the version up in package.json!", "error");
} else {
  let component = null;
  let style = null;

  glob(buildFilePattern, function (er, files) {
    if (files.length > 0) {
      component = files[0];

      glob(buildStylesPattern, function (er, files) {
        if (files.length > 0) {
          style = files[0];

          if (component == null || style == null) {
            log("Could not locate one or more build soure files!", "error");
          } else {
            const dirExists = fs.existsSync(releaseDirPath);
            if (!dirExists) {
              fs.mkdir(releaseDirPath, { recursive: true }, (err) => {
                if (err) {
                  log(`Failed to initialize the release dir: ${releaseDirPath}`, "error");
                }
              });
              triggerRelease(component, releaseFilePath, style, releaseStylesPath, releaseHTMLPath);
            }
          }
        } else {
          log("Could not locate one or more build soure files!", "error");
        }
      });
    } else {
      log("Could not locate one or more build soure files!", "error");
    }
  });
}

function triggerRelease(component, releaseFilePath, style, releaseStylesPath, releaseHTMLPath) {
  fs.copyFile(component, releaseFilePath, (err) => {
    if (err) {
      log("Releasing Component Failed", "error");
    } else {
      log(`Released a new component version: (${path.basename(releaseFilePath)}). You may find it in the ./release dir.`, "success");
    }
  });

  fs.copyFile(style, releaseStylesPath, (err) => {
    if (err) {
      log("Releasing Styles Failed", "error");
    } else {
      log(`Released a new Styles version: (${path.basename(releaseStylesPath)}). You may find it in the ./release dir.`, "success");
    }
  });

  let testHtml = `
  <!DOCTYPE html>
  <html lang="en">

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="theme-color" content="#000000" />
    <meta name="description" content="Web site created using create-react-app" />
    <title>Kolloqe | Socketio Widget</title>
  </head>

  <body>
    <h2>Kolloqe Socketio Chat Widget</h2>
    <div id="testid"></div>

    <noscript>You need to enable JavaScript to run this app.</noscript>
    <script>
      window.kolloqe_webchat_id = "testid";
      window.kolloqeProps = {
        testid: {
          position: "right",
          positionProps: {
            xAxis: "10px",
            yAxis: "10px"
          },
          title: "Kolloqe ⚡",
          subtitle: "chat widget",
          defaultLang: "en",
          avatar: true,
          avatarType: "text",
          avatarLink: "",
          socketURL: "http://localhost:5005",
          initPayload: "/get_started",
          reconnectTimeout: 5000,
          widgetOptions: true,
          fullscreenButton: false,
          langSwitch: true,
          moreOptions: true,
          widgetOptionsPosition: "top",
          hideWhenNotConnected: true,
          displayUnreadCount: false,
          showMessageDate: false,
          customMessageDelay: null,
          persistSession: true,
          triggerIntents: true,
          enableURLs: true,
          disablePreviousQuickReplies: true,
          indicatorDelay: 0,
          indicatorType: "bouncing",
          params: null,
          enableShortcuts: true,
          shortcutKey: 81,
        }
      };

      const version = "${packageJSON.version}";
      const source = "kolloqe-widget-socketio-en-html-v${packageJSON.version}";

      !(function () {
        let e = document.createElement("script");
        let s = document.createElement("link");
        let t = document.head || document.getElementsByTagName("head")[0];
        (e.src = "kolloqe-widget-socketio-en-html-v${packageJSON.version}.js"),
          (e.async = !0),
          (e.onload = () => { }),
          t.insertBefore(e, t.firstChild);

        (s.href = "kolloqe-widget-socketio-en-html-v${packageJSON.version}.css"),
          (s.rel = "stylesheet"),
          (s.type = "text/css"),
          t.insertBefore(s, t.firstChild);
      })();
    </script>
  </body>

  </html>
  `;

  fs.writeFile(releaseHTMLPath, testHtml, (error) => {
    if (error) {
      log("Failed to generate the test HTML page. You might need to build it manually!", "error");
    }
    log(`Released a test HTML: (${path.basename(releaseHTMLPath)}). You may find it in the ./release dir.`, "success");
  });
}

function log(text, type) {
  if (type === "error") {
    console.error(
      "\x1b[31m",
      text,
      "\x1b[0m"
    );
    process.exit();
  } else if (type === "success") {
    console.log(
      "\x1b[32m",
      text,
      "\x1b[0m"
    );
  } else {
    console.log(text);
  }
}